torchao/quantization/quantize_/workflows/int8/int8_tensor.py

```diff
@@ -0,0 +1,303 @@
+            f"block_size={self.block_size}, "
+            f"shape={self.shape}, "
+            f"device={self.device}, "
+            f"dtype={self.dtype})"
+        )
+
+    @classmethod
+    def from_hp(
+        cls,
+        w_hp: torch.Tensor,
+        granularity: Granularity = PerRow(),
+        act_quant_kwargs: Optional[QuantizeTensorToInt8Kwargs] = None,
+    ):
+        """Create Int8Tensor from high-precision tensor"""
+        block_size = get_block_size(w_hp.shape, granularity)
+
+        if w_hp.dim() not in [2, 3] or len(block_size) != w_hp.dim():
+            raise ValueError(
+                f"Expected 2D or 3D tensor with matching block_size dimensions, "
+                f"got tensor dim={w_hp.dim()}, block_size length={len(block_size)}"
+            )
+
+        scale, zero_point = choose_qparams_affine(
+            input=w_hp,
+            mapping_type=MappingType.SYMMETRIC,
+            block_size=block_size,
+            target_dtype=torch.int8,
+            quant_min=-128,
+            quant_max=127,
+            scale_dtype=w_hp.dtype,
+            zero_point_dtype=torch.int8,
+        )
+
+        int_data = quantize_affine(
+            w_hp,
+            block_size=block_size,
+            scale=scale,
+            zero_point=zero_point,
+            output_dtype=torch.int8,
+        )
+
+        return cls(
+            int_data,
+            scale,
+            granularity,
+            block_size=block_size,
+            act_quant_kwargs=act_quant_kwargs,
+            dtype=w_hp.dtype,
+        )
+
+    def dequantize(self, output_dtype: Optional[torch.dtype] = None) -> torch.Tensor:
+        """Dequantize int8 tensor to floating point"""
+        if output_dtype is None:
+            output_dtype = self.dtype
+
+        block_size = get_block_size(self.qdata.shape, self.granularity)
+
+        return dequantize_affine(
+            input=self.qdata,
+            block_size=block_size,
+            scale=self.scale,
+            zero_point=None,
+            input_dtype=torch.int8,
+            quant_min=-128,
+            quant_max=127,
+            output_dtype=output_dtype,
+        )
+
+
+implements = Int8Tensor.implements
+implements_torch_function = Int8Tensor.implements_torch_function
+
+
+@implements(aten.linear.default)
+@implements_torch_function(torch.nn.functional.linear)
+def _(func, types, args, kwargs):
+    """INT8 quantization: dynamic activation or weight-only"""
+    activation_tensor, weight_tensor, bias = (
+        args[0],
+        args[1],
+        args[2] if len(args) > 2 else None,
+    )
+
+    if not isinstance(weight_tensor, Int8Tensor):
+        raise TypeError(f"Expected weight to be Int8Tensor, got {type(weight_tensor)}")
+
+    output_dtype = activation_tensor.dtype
+
+    if weight_tensor.act_quant_kwargs is not None:
+        activation_tensor = Int8Tensor.from_hp(
+            activation_tensor, weight_tensor.act_quant_kwargs.granularity
+        )
```

Comment:
this is correct, btw I didn't mean to remove these from https://github.com/pytorch/ao/blob/2ff1eb2e356275cfbe46832327387d382c72945d/torchao/quantization/quantize_/common/quantize_tensor_kwargs.py#L33. what I meant was just not pass act_quant_kwargs again there, but anyways we can fix later